src/Calf/Parallel.agda

```diff
@@ -0,0 +1,74 @@
+{-# OPTIONS --rewriting --allow-unsolved-metas #-}
+
+-- The basic CBPV metalanguage, extended with parallelism.
+
+open import Algebra.Cost
+
+module Calf.Parallel (parCostMonoid : ParCostMonoid) where
+
+open ParCostMonoid parCostMonoid
+
+open import Calf.Prelude
+open import Calf.CBPV
+open import Calf.Directed
+open import Calf.Step costMonoid
+
+open import Calf.Data.Product
+open import Calf.Data.IsBoundedG costMonoid
+open import Calf.Data.IsBounded costMonoid
+
+open import Data.Product
+open import Relation.Binary.PropositionalEquality
+
+postulate
+  _∥_ : {A₁ A₂ : tp pos} → cmp (F A₁) → cmp (F A₂) → cmp (F (A₁ ×⁺ A₂))
+
+  ∥/join : ∀ {A₁ A₂} {v₁ v₂ c₁ c₂} →
+    step (F A₁) c₁ (ret v₁) ∥ step (F A₂) c₂ (ret v₂) ≡ step (F (A₁ ×⁺ A₂)) (c₁ ⊗ c₂) (ret (v₁ , v₂))
+
+∥/join/𝟘 : ∀ {A₁ A₂} {v₁ : val A₁} {v₂ : val A₂} →
+  ret {A₁} v₁ ∥ ret {A₂} v₂ ≡ ret (v₁ , v₂)
+∥/join/𝟘 {A₁} {A₂} {v₁} {v₂} =
+  let open ≡-Reasoning in
+  begin
+    ret v₁ ∥ ret v₂
+  ≡⟨⟩
+    step (F A₁) 𝟘 (ret v₁) ∥ step (F A₂) 𝟘 (ret v₂)
+  ≡⟨ ∥/join {A₁} {A₂} {v₁} {v₂} {𝟘} {𝟘} ⟩
+    step (F (Σ⁺ A₁ (λ _ → A₂))) (𝟘 ⊗ 𝟘) (ret (v₁ , v₂))
+  ≡⟨ cong (λ c → step (F (Σ⁺ A₁ (λ _ → A₂))) c (ret (v₁ , v₂))) (⊗-identityˡ 𝟘) ⟩
+    step (F (Σ⁺ A₁ (λ _ → A₂))) 𝟘 (ret (v₁ , v₂))
+  ≡⟨⟩
+    ret (v₁ , v₂)
+  ∎
+{-# REWRITE ∥/join ∥/join/𝟘 #-}
+
+∥-mono-≲ : {A₁ A₂ : tp pos} {e₁ e₁' : cmp (F A₁)} {e₂ e₂' : cmp (F A₂)}
+  → e₁ ≤⁻[ F A₁ ] e₁'
+  → e₂ ≤⁻[ F A₂ ] e₂'
+  → (e₁ ∥ e₂) ≤⁻[ F (A₁ ×⁺ A₂) ] (e₁' ∥ e₂')
+∥-mono-≲ {A₁} {A₂} {e₁} {e₁'} {e₂} {e₂'} e₁≤e₁' e₂≤e₂' =
+  ≤⁻-mono₂ _∥_ e₁≤e₁' e₂≤e₂'
+
+boundg/par : {A₁ A₂ : tp pos} {e₁ : cmp (F A₁)} {e₂ : cmp (F A₂)} {b₁ b₂ : cmp cost} →
+  IsBoundedG A₁ e₁ b₁ →
+  IsBoundedG A₂ e₂ b₂ →
+  IsBoundedG (Σ⁺ A₁ λ _ → A₂) (e₁ ∥ e₂) (bind cost (b₁ ∥ b₂) λ _ → ret triv)
+boundg/par {A₁} {A₂} {e₁} {e₂} {b₁} {b₂} ib₁ ib₂ =
+  let open ≤⁻-Reasoning cost in
+  begin
+    bind cost (e₁ ∥ e₂) (λ _ → ret triv)
+  ≤⟨ {!   !} ⟩
```

Comment:
@kaonn You mentioned that you think you can fill this hole: https://github.com/jonsterling/agda-calf/commit/beb21724e2fe74aeaaaabf5a49e9656c3b7e9327#r121039400

I still don't quite see how this would work; do you think you'd be able to fill it?